indico/modules/auth/providers.py

```diff
@@ -30,6 +31,25 @@ def check_password(self, identity, password):
             session['insecure_password_error'] = error
         return True
 
+    def process_local_login(self, data):
+        identity = (Identity.query.filter(Identity.provider == self.name,
+                                          db.or_(Identity.identifier == data['identifier'],
+                                                 UserEmail.email == data['identifier']),
+                                          ~User.is_pending,
+                                          ~UserEmail.is_user_deleted)
+                                  .join(User)
+                                  .join(UserEmail).first())
```

Comment:
I don't think this is always correct. You query a single identity but I think we should consider edge cases where you have an existing username that looks like an email address.

For example, on my test instance I have an account with `a.m@cern.ch` and then created a local account with a username of `a.m@cern.ch`.

Now this query just gives me one identity, which may or may not be the correct one. That's why we need to use `.all()` here and iterate over the matched identities:

- If there are no identities --> NoSuchUser
- If the password check fails for all of them --> InvalidCredentials
- otherwise continue with the success logic for the first identity where the password was correct (I think we can safely ignore the case where the password is also correct for multiple identities, in that case it's almost certainly the same person just doing something dumb w/ multiple accounts and they'll probably open a support case eventually leading to merging accounts)

So basically this:

```py
    def process_local_login(self, data):
        # Query identities either by (user-settable) username or by email address.
        # In some very rare edge cases we can get more than one, e.g. when a user has an old identity
        # with a username that happens to be the email address of another user (this is no longer possible,
        # but existing accounts could have it).
        identities = (
            Identity.query
            .join(User)
            .join(UserEmail)
            .filter(Identity.provider == self.name,
                    db.or_(Identity.identifier == data['identifier'],
                           UserEmail.email == data['identifier']),
                    ~User.is_pending,
                    ~UserEmail.is_user_deleted)
            .order_by(Identity.id)
            .all()
        )
        # If we have no matching identities we fail with invalid username
        # XXX This is intentional, having an account on an Indico instance is generally not considered
        # secret information, and we want to give users the benefit of more verbose error messages.
        if not identities:
            raise NoSuchUser(provider=self)
        # From all the matching identities (usually just one), get one where the password matches, or
        # fail with invalid-password if there is none.
        if not (identity := next((ide for ide in identities if self.check_password(ide, data['password'])), None)):
            raise InvalidCredentials(provider=self)
        if data['identifier'] != identity.identifier and data['identifier'] in identity.user.secondary_emails:
            flash(_('You are logging in with a secondary email address. Please note that account related '
                    'notifications will always be sent to your primary email address. Go to "My profile" to '
                    'manage your emails.'), 'warning')
        auth_info = AuthInfo(self, identity=identity)
        return self.multipass.handle_auth_success(auth_info)
```

Feel free to copy this.

Reply:
Done